lua/weapons/weapon_ttt_minty_blink/shared.lua

```diff
@@ -0,0 +1,471 @@
+    self.Timer.Particle = particle and CurTime() or 0
+    self.Timer.Recharge = recharge and CurTime() or 0
+    self.Timer.PostProcess = post_process and CurTime() or 0
+end
+
+function SWEP:Blink_DoViewModelAnimation()
+    local owner = self.Owner
+    if not IsValid( owner ) then return end
+
+    local viewmodel = owner:GetViewModel()
+    if not IsValid( viewmodel ) then return end
+
+    if SERVER or IsFirstTimePredicted() then
+        viewmodel:SendViewModelMatchingSequence( 12 )
+    end
+end
+
+function SWEP:Blink_DoAimTrace()
+    local owner = self.Owner
+    if not IsValid( owner ) then return nil end
+
+    local trace_hull = Vector( trace_hull_size, trace_hull_size, trace_hull_size )
+
+    -- Initial trace
+    local marker_trace = util.TraceLine( {
+        start   = owner:EyePos() + owner:GetAimVector(),
+        endpos  = owner:EyePos() + owner:GetAimVector() * cvar.range:GetInt(),
+        mask    = MASK_ALL,
+        filter  = owner
+    } )
+
+    -- Find ledge
+    local marker_hit_wall = marker_trace.Hit and math.abs( marker_trace.HitNormal.z ) < 0.15
+    local ledge_trace = nil
+
+    if marker_hit_wall then
+        for i = 1, ledge_step_max do
+            local origin = marker_trace.HitPos + Vector( 0, 0, i * ledge_step_size )
+            
+            local candidate = util.TraceLine( {
+                start   = origin,
+                endpos  = origin - ( marker_trace.HitNormal * trace_ledge_depth ),
+                mask    = MASK_PLAYERSOLID,
+                filter  = owner
+            } )
+
+            dbg.ledge_check( candidate )
+
+            if not candidate.Hit then
+                ledge_trace = candidate
+                break
+            end
+        end
+    end
+
+    -- Find ground
+    local ground_trace_origin =
+        ledge_trace and ledge_trace.HitPos or 
+        marker_hit_wall and ( marker_trace.HitPos + marker_trace.HitNormal * trace_hull_size * 2 ) or
+        marker_trace.HitPos
+
+    local ground_trace = util.TraceLine( {
+        start   = ground_trace_origin,
+        endpos  = ground_trace_origin - Vector( 0, 0, 65535 ),
+        mask    = MASK_PLAYERSOLID,
+        filter  = owner
+    } )
+
+    ground_trace = util.TraceHull( { -- Detect highest point on uneven surface
+        start   = ground_trace.HitPos - ( ground_trace.Normal * trace_hull_size ),
+        endpos  = ground_trace.HitPos,
+        mins    = -trace_hull,
+        maxs    = trace_hull,
+        mask    = MASK_PLAYERSOLID,
+        filter  = owner
+    } )
+
+    -- Ensure player fit
+    local player_mins, player_maxs = owner:GetCollisionBounds()
+    local final_trace = nil
+
+    local origin = ground_trace.HitPos
+    local normal = ledge_trace and ledge_trace.Normal or marker_trace.HitNormal
+    local angle = normal:Angle()
+
+    local use_lut =
+        ledge_trace and lut.ledge or 
+        marker_hit_wall and lut.wall or
+        lut.floor
+
+    for k, v in pairs( use_lut ) do -- Evaluate points
```

Comment:
You can use `ipairs` instead here since all three of the lut tables use sequential keys (ipairs is faster than pairs)